illumos-utils/src/lib.rs

```diff
@@ -108,6 +111,33 @@ mod inner {
     }
 }
 
+// Due to feature unification, the `testing` feature is enabled when some tests
+// don't actually want to use it. We allow them to opt out of  the use of the
+// free function here. We also explicitly opt-in where mocks are used.
+//
+// Note that this only works if the tests that use mocks and those that  don't
+// are run sequentially. However, this is how we do things in CI with nextest,
+// so there is no problem currently.
+//
+// We can remove all this when we get rid of the mocks.
+pub static USE_MOCKS: AtomicBool = AtomicBool::new(false);
```

Comment:
I definitely understand your concern here. I felt super gross when writing this. However, it took me many hours to even come up with this solution after trying every other thing I could think of. This was the only way I could come up with to allow mocking of functions to be used in some crates and not other crates when `cargo nextest run` at the top level was used. 

The problem is that as opposed to the mocking of structs and enums, we don't distinguish by names in tests the mocked version of the methods, and so we can't choose whether to use them or not. So without this, when I ran `cargo nextest -p sled-storage` it would pass, but when run with all the other tests with `cargo nextest run` the tests would fail because I'd be missing a mock for `execute`, even though I didn't actually want one!

It's of note that this was also a problem when converting over the tests inside sled-agent for zone-bundler to use the real illumos crate code and temp directories as was [originally desired](https://github.com/oxidecomputer/omicron/blob/c8db7c88dec4030f29bea0520dea43e81187c8dd/sled-agent/src/zone_bundle.rs#L2202-L2218). I ended up reverting that to how the code exists on main now to avoid this  issue.

AFAICT the only place where these mocks are used are in the [ServiceManager tests](https://github.com/oxidecomputer/omicron/blob/main/sled-agent/src/services.rs#L3080-L3142), and of those expectations, the only one that interferes is `execute_ctx` because of the module wrapped function compile time choice and feature unification shenanigans. I think the way forward is to do a similar refactor of the ServiceManager so that tis functionality can be tested without mocks. Then we can remove all the mock stuff in `illumos_utils`. 

I also believe your abstraction related to tokamak does this, and we could also get this merged in. In either case, getting rid of the mocks is where the win is.

I will go ahead and create an issue here. I was able to go ahead and make it so that at least this static doesn't exist in production builds in 788421384 so that should alleviate some of the concern at I hope.

Reply:
Issue Created: https://github.com/oxidecomputer/omicron/issues/4481

Reply:
Hiding this behind a cfg helps a lot, thank you!